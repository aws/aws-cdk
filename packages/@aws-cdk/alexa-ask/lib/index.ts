// Alexa::ASK CloudFormation Resources:
export * from './ask.generated';
export * from './pipeline-actions';
