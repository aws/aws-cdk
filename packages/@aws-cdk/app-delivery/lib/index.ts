export * from './pipeline-deploy-stack-action';
