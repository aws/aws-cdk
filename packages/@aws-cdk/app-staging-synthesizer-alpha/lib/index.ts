export * from './default-staging-stack';
export * from './app-staging-synthesizer';
export * from './bootstrap-roles';
export * from './staging-stack';
