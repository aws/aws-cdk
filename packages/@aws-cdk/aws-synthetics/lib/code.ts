import * as fs from 'fs';
import * as path from 'path';
import * as s3 from '@aws-cdk/aws-s3';
import * as s3_assets from '@aws-cdk/aws-s3-assets';
import { Construct } from '@aws-cdk/core';

/**
 * The code the canary should execute
 */
export abstract class Code {

  /**
   * Specify code inline.
   *
   * @param code The actual handler code (limited to 4KiB)
   *
   * @returns `InlineCode` with inline code.
   */
  public static fromInline(code: string): InlineCode {
    return new InlineCode(code);
  }

  /**
   * Specify code from a local path. Path must include the folder structure `nodejs/node_modules/myCanaryFilename.js`.
   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries_WritingCanary.html#CloudWatch_Synthetics_Canaries_write_from_scratch
   *
   * @param assetPath Either a directory or a .zip file
   *
   * @returns `AssetCode` associated with the specified path.
   */
  public static fromAsset(assetPath: string, options?: s3_assets.AssetOptions): AssetCode {
    return new AssetCode(assetPath, options);
  }

  /**
   * Specify code from an s3 bucket. The object in the s3 bucket must be a .zip file that contains
   * the structure `nodejs/node_modules/myCanaryFilename.js`.
   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries_WritingCanary.html#CloudWatch_Synthetics_Canaries_write_from_scratch
   *
   * @param bucket The S3 bucket
   * @param key The object key
   * @param objectVersion Optional S3 object version
   *
   * @returns `S3Code` associated with the specified S3 object.
   */
  public static fromBucket(bucket: s3.IBucket, key: string, objectVersion?: string): S3Code {
    return new S3Code(bucket, key, objectVersion);
  }

  /**
   * Called when the canary is initialized to allow this object to bind
   * to the stack, add resources and have fun.
   *
   * @param scope The binding scope. Don't be smart about trying to down-cast or
   *              assume it's initialized. You may just use it as a construct scope.
   *
   * @returns a bound `CodeConfig`.
   */
  public abstract bind(scope: Construct, handler: string): CodeConfig;
}

/**
 * Configuration of the code class
 */
export interface CodeConfig {
  /**
   * The location of the code in S3 (mutually exclusive with `inlineCode`).
   *
   * @default - none
   */
  readonly s3Location?: s3.Location;

  /**
   * Inline code (mutually exclusive with `s3Location`).
   *
   * @default - none
   */
  readonly inlineCode?: string;
}

/**
 * Canary code from an Asset
 */
export class AssetCode extends Code {
  private asset?: s3_assets.Asset;

  /**
   * @param assetPath The path to the asset file or directory.
   */
  public constructor(private assetPath: string, private options?: s3_assets.AssetOptions) {
    super();

    if (!fs.existsSync(this.assetPath)) {
      throw new Error(`${this.assetPath} is not a valid path`);
    }
  }

  public bind(scope: Construct, handler: string): CodeConfig {
    this.validateCanaryAsset(handler);

    // If the same AssetCode is used multiple times, retain only the first instantiation.
    if (!this.asset) {
      this.asset = new s3_assets.Asset(scope, 'Code', {
        path: this.assetPath,
        ...this.options,
      });
    }

    return {
      s3Location: {
        bucketName: this.asset.s3BucketName,
        objectKey: this.asset.s3ObjectKey,
      },
    };
  }

  /**
   * Validates requirements specified by the canary resource. For example, the canary code with handler `index.handler`
   * must be found in the file structure `nodejs/node_modules/index.js`.
   *
   * Requires path to be either zip file or directory.
   * Requires asset directory to have the structure 'nodejs/node_modules'.
   * Requires canary file to be directly inside node_modules folder.
   * Requires canary file name matches the handler name.
   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries_WritingCanary.html
   *
   * @param handler the canary handler
   */
  private validateCanaryAsset(handler: string) {
    if (path.extname(this.assetPath) !== '.zip') {
      if (!fs.lstatSync(this.assetPath).isDirectory()) {
        throw new Error(`Asset must be a .zip file or a directory (${this.assetPath})`);
      }
      const filename = `${handler.split('.')[0]}.js`;
      if (!fs.existsSync(path.join(this.assetPath, 'nodejs', 'node_modules', filename))) {
        throw new Error(`The canary resource requires that the handler is present at "nodejs/node_modules/${filename}" but not found at ${this.assetPath} (https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries_WritingCanary.html#CloudWatch_Synthetics_Canaries_write_from_scratch)`);
      }
    }
  }
}

/**
 * Canary code from an inline string.
 */
export class InlineCode extends Code {
  public constructor(private code: string) {
    super();

    if (code.length === 0) {
      throw new Error('Canary inline code cannot be empty');
    }
  }

  public bind(_scope: Construct, handler: string): CodeConfig {

    if (handler !== 'index.handler') {
      throw new Error(`The handler for inline code must be "index.handler" (got "${handler}")`);
    }

    return {
      inlineCode: this.code,
    };
  }
}

/**
 * S3 bucket path to the code zip file
 */
export class S3Code extends Code {
  public constructor(private bucket: s3.IBucket, private key: string, private objectVersion?: string) {
    super();
  }

  public bind(_scope: Construct, _handler: string): CodeConfig {
    return {
      s3Location: {
        bucketName: this.bucket.bucketName,
        objectKey: this.key,
        objectVersion: this.objectVersion,
      },
    };
  }
}
