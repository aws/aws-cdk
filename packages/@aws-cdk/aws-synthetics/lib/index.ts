export * from './canary';
export * from './code';
export * from './schedule';

// AWS::Synthetics CloudFormation Resources:
export * from './synthetics.generated';
