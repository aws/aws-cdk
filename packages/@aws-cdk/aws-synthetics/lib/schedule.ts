import { Duration } from '@aws-cdk/core';

/**
 * Schedule for canary runs
 */
export class Schedule {

  /**
   * The canary will be executed once.
   */
  public static once(): Schedule {
    return new Schedule('rate(0 minutes)');
  }

  /**
   * Construct a schedule from a literal schedule expression. The expression must be in a `rate(number units)` format.
   * For example, `Schedule.expression('rate(10 minutes)')`
   *
   * @param expression The expression to use.
   */
  public static expression(expression: string): Schedule {
    return new Schedule(expression);
  }

  /**
   * Construct a schedule from an interval. Allowed values: 0 (for a single run) or between 1 and 60 minutes.
   * To specify a single run, you can use `Schedule.once()`.
   *
   * @param interval The interval at which to run the canary
   */
  public static rate(interval: Duration): Schedule {
    const minutes = interval.toMinutes();
    if (minutes > 60) {
      throw new Error('Schedule duration must be between 1 and 60 minutes');
    }
    if (minutes === 0) {
      return Schedule.once();
    }
    if (minutes === 1) {
      return new Schedule('rate(1 minute)');
    }
    return new Schedule(`rate(${minutes} minutes)`);
  }

  private constructor(
    /**
     * The Schedule expression
     */
    public readonly expressionString: string) {}
}
