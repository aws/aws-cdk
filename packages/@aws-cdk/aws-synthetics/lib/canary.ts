import * as crypto from 'crypto';
import { Metric, MetricOptions, MetricProps } from '@aws-cdk/aws-cloudwatch';
import * as iam from '@aws-cdk/aws-iam';
import * as s3 from '@aws-cdk/aws-s3';
import * as cdk from '@aws-cdk/core';
import { Construct } from 'constructs';
import { Code } from './code';
import { Schedule } from './schedule';
import { CloudWatchSyntheticsMetrics } from './synthetics-canned-metrics.generated';
import { CfnCanary } from './synthetics.generated';

/**
 * Specify a test that the canary should run
 */
export class Test {
  /**
   * Specify a custom test with your own code
   *
   * @returns `Test` associated with the specified Code object
   * @param options The configuration options
   */
  public static custom(options: CustomTestOptions): Test {
    Test.validateHandler(options.handler);
    return new Test(options.code, options.handler);
  }

  /**
   * Verifies that the given handler ends in '.handler'. Returns the handler if successful and
   * throws an error if not.
   *
   * @param handler - the handler given by the user
   */
  private static validateHandler(handler: string) {
    if (!handler.endsWith('.handler')) {
      throw new Error(`Canary Handler must end in '.handler' (${handler})`);
    }
    if (handler.length > 21) {
      throw new Error(`Canary Handler must be less than 21 characters (${handler})`);
    }
  }

  /**
   * Construct a Test property
   *
   * @param code The code that the canary should run
   * @param handler The handler of the canary
   */
  private constructor(public readonly code: Code, public readonly handler: string) {
  }
}

/**
 * Properties for specifying a test
 */
export interface CustomTestOptions {
  /**
   * The code of the canary script
   */
  readonly code: Code,

  /**
   * The handler for the code. Must end with `.handler`.
   */
  readonly handler: string,
}

/**
 * Runtime options for a canary
 */
export class Runtime {
  /**
   * `syn-1.0` includes the following:
   *
   * - Synthetics library 1.0
   * - Synthetics handler code 1.0
   * - Lambda runtime Node.js 10.x
   * - Puppeteer-core version 1.14.0
   * - The Chromium version that matches Puppeteer-core 1.14.0
   *
   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Library_nodejs_puppeteer.html#CloudWatch_Synthetics_runtimeversion-1.0
   */
  public static readonly SYNTHETICS_1_0 = new Runtime('syn-1.0');

  /**
   * `syn-nodejs-2.0` includes the following:
   * - Lambda runtime Node.js 10.x
   * - Puppeteer-core version 3.3.0
   * - Chromium version 83.0.4103.0
   *
   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Library_nodejs_puppeteer.html#CloudWatch_Synthetics_runtimeversion-2.0
   */
  public static readonly SYNTHETICS_NODEJS_2_0 = new Runtime('syn-nodejs-2.0');


  /**
   * `syn-nodejs-2.1` includes the following:
   * - Lambda runtime Node.js 10.x
   * - Puppeteer-core version 3.3.0
   * - Chromium version 83.0.4103.0
   *
   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Library_nodejs_puppeteer.html#CloudWatch_Synthetics_runtimeversion-2.1
   */
  public static readonly SYNTHETICS_NODEJS_2_1 = new Runtime('syn-nodejs-2.1');

  /**
   * `syn-nodejs-2.2` includes the following:
   * - Lambda runtime Node.js 10.x
   * - Puppeteer-core version 3.3.0
   * - Chromium version 83.0.4103.0
   *
   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Library_nodejs_puppeteer.html#CloudWatch_Synthetics_runtimeversion-2.2
   */
  public static readonly SYNTHETICS_NODEJS_2_2 = new Runtime('syn-nodejs-2.2');

  /**
   * `syn-nodejs-puppeteer-3.0` includes the following:
   * - Lambda runtime Node.js 12.x
   * - Puppeteer-core version 5.5.0
   * - Chromium version 88.0.4298.0
   *
   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Library_nodejs_puppeteer.html#CloudWatch_Synthetics_runtimeversion-nodejs-puppeteer-3.0
   */
  public static readonly SYNTHETICS_NODEJS_PUPPETEER_3_0 = new Runtime('syn-nodejs-puppeteer-3.0');

  /**
   * `syn-nodejs-puppeteer-3.1` includes the following:
   * - Lambda runtime Node.js 12.x
   * - Puppeteer-core version 5.5.0
   * - Chromium version 88.0.4298.0
   *
   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Library_nodejs_puppeteer.html#CloudWatch_Synthetics_runtimeversion-nodejs-puppeteer-3.1
   */
  public static readonly SYNTHETICS_NODEJS_PUPPETEER_3_1 = new Runtime('syn-nodejs-puppeteer-3.1');

  /**
   * @param name The name of the runtime version
   */
  public constructor(public readonly name: string) {
  }
}

/**
 * Options for specifying the s3 location that stores the data of each canary run. The artifacts bucket location **cannot**
 * be updated once the canary is created.
 */
export interface ArtifactsBucketLocation {
  /**
   * The s3 location that stores the data of each run.
   */
  readonly bucket: s3.IBucket;

  /**
   * The S3 bucket prefix. Specify this if you want a more specific path within the artifacts bucket.
   *
   * @default - no prefix
   */
  readonly prefix?: string;
}

/**
 * Properties for a canary
 */
export interface CanaryProps {
  /**
   * The s3 location that stores the data of the canary runs.
   *
   * @default - A new s3 bucket will be created without a prefix.
   */
  readonly artifactsBucketLocation?: ArtifactsBucketLocation;

  /**
   * Canary execution role.
   *
   * This is the role that will be assumed by the canary upon execution.
   * It controls the permissions that the canary will have. The role must
   * be assumable by the AWS Lambda service principal.
   *
   * If not supplied, a role will be created with all the required permissions.
   * If you provide a Role, you must add the required permissions.
   *
   * @see required permissions: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-synthetics-canary.html#cfn-synthetics-canary-executionrolearn
   *
   * @default - A unique role will be generated for this canary.
   * You can add permissions to roles by calling 'addToRolePolicy'.
   */
  readonly role?: iam.IRole;

  /**
   * How long the canary will be in a 'RUNNING' state. For example, if you set `timeToLive` to be 1 hour and `schedule` to be `rate(10 minutes)`,
   * your canary will run at 10 minute intervals for an hour, for a total of 6 times.
   *
   * @default - no limit
   */
  readonly timeToLive?: cdk.Duration;

  /**
   * Specify the schedule for how often the canary runs. For example, if you set `schedule` to `rate(10 minutes)`, then the canary will run every 10 minutes.
   * You can set the schedule with `Schedule.rate(Duration)` (recommended) or you can specify an expression using `Schedule.expression()`.
   * @default 'rate(5 minutes)'
   */
  readonly schedule?: Schedule;

  /**
   * Whether or not the canary should start after creation.
   *
   * @default true
   */
  readonly startAfterCreation?: boolean;

  /**
   * How many days should successful runs be retained.
   *
   * @default Duration.days(31)
   */
  readonly successRetentionPeriod?: cdk.Duration;

  /**
   * How many days should failed runs be retained.
   *
   * @default Duration.days(31)
   */
  readonly failureRetentionPeriod?: cdk.Duration;

  /**
   * The name of the canary. Be sure to give it a descriptive name that distinguishes it from
   * other canaries in your account.
   *
   * Do not include secrets or proprietary information in your canary name. The canary name
   * makes up part of the canary ARN, which is included in outbound calls over the internet.
   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/servicelens_canaries_security.html
   *
   * @default - A unique name will be generated from the construct ID
   */
  readonly canaryName?: string;

  /**
   * Specify the runtime version to use for the canary.
   *
   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries_Library.html
   */
  readonly runtime: Runtime;

  /**
   * The type of test that you want your canary to run. Use `Test.custom()` to specify the test to run.
   */
  readonly test: Test;

  /**
   * Key-value pairs that the Synthetics caches and makes available for your canary scripts. Use environment variables
   * to apply configuration changes, such as test and production environment configurations, without changing your
   * Canary script source code.
   *
   * @default - No environment variables.
   */
  readonly environmentVariables?: { [key: string]: string };
}

/**
 * Define a new Canary
 */
export class Canary extends cdk.Resource {
  /**
   * Execution role associated with this Canary.
   */
  public readonly role: iam.IRole;

  /**
   * The canary ID
   * @attribute
   */
  public readonly canaryId: string;

  /**
   * The state of the canary. For example, 'RUNNING', 'STOPPED', 'NOT STARTED', or 'ERROR'.
   * @attribute
   */
  public readonly canaryState: string;

  /**
   * The canary Name
   * @attribute
   */
  public readonly canaryName: string;

  /**
   * Bucket where data from each canary run is stored.
   */
  public readonly artifactsBucket: s3.IBucket;

  public constructor(scope: Construct, id: string, props: CanaryProps) {
    if (props.canaryName && !cdk.Token.isUnresolved(props.canaryName)) {
      validateName(props.canaryName);
    }

    super(scope, id, {
      physicalName: props.canaryName || cdk.Lazy.string({
        produce: () => this.generateUniqueName(),
      }),
    });

    this.artifactsBucket = props.artifactsBucketLocation?.bucket ?? new s3.Bucket(this, 'ArtifactsBucket', {
      encryption: s3.BucketEncryption.KMS_MANAGED,
    });

    this.role = props.role ?? this.createDefaultRole(props.artifactsBucketLocation?.prefix);

    const resource: CfnCanary = new CfnCanary(this, 'Resource', {
      artifactS3Location: this.artifactsBucket.s3UrlForObject(props.artifactsBucketLocation?.prefix),
      executionRoleArn: this.role.roleArn,
      startCanaryAfterCreation: props.startAfterCreation ?? true,
      runtimeVersion: props.runtime.name,
      name: this.physicalName,
      schedule: this.createSchedule(props),
      failureRetentionPeriod: props.failureRetentionPeriod?.toDays(),
      successRetentionPeriod: props.successRetentionPeriod?.toDays(),
      code: this.createCode(props),
      runConfig: this.createRunConfig(props),
    });

    this.canaryId = resource.attrId;
    this.canaryState = resource.attrState;
    this.canaryName = this.getResourceNameAttribute(resource.ref);
  }

  /**
   * Measure the Duration of a single canary run, in seconds.
   *
   * @param options - configuration options for the metric
   *
   * @default avg over 5 minutes
   */
  public metricDuration(options?: MetricOptions): Metric {
    return this.cannedMetric(CloudWatchSyntheticsMetrics.durationAverage, options);
  }

  /**
   * Measure the percentage of successful canary runs.
   *
   * @param options - configuration options for the metric
   *
   * @default avg over 5 minutes
   */
  public metricSuccessPercent(options?: MetricOptions): Metric {
    return this.cannedMetric(CloudWatchSyntheticsMetrics.successPercentAverage, options);
  }

  /**
   * Measure the number of failed canary runs over a given time period.
   *
   * Default: sum over 5 minutes
   *
   * @param options - configuration options for the metric
   */
  public metricFailed(options?: MetricOptions): Metric {
    return this.cannedMetric(CloudWatchSyntheticsMetrics.failedSum, options);
  }

  /**
   * Returns a default role for the canary
   */
  private createDefaultRole(prefix?: string): iam.IRole {
    const { partition } = cdk.Stack.of(this);
    // Created role will need these policies to run the Canary.
    // https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-synthetics-canary.html#cfn-synthetics-canary-executionrolearn
    const policy = new iam.PolicyDocument({
      statements: [
        new iam.PolicyStatement({
          resources: ['*'],
          actions: ['s3:ListAllMyBuckets'],
        }),
        new iam.PolicyStatement({
          resources: [this.artifactsBucket.arnForObjects(`${prefix ? prefix+'/*' : '*'}`)],
          actions: ['s3:PutObject', 's3:GetBucketLocation'],
        }),
        new iam.PolicyStatement({
          resources: ['*'],
          actions: ['cloudwatch:PutMetricData'],
          conditions: { StringEquals: { 'cloudwatch:namespace': 'CloudWatchSynthetics' } },
        }),
        new iam.PolicyStatement({
          resources: [`arn:${partition}:logs:::*`],
          actions: ['logs:CreateLogStream', 'logs:CreateLogGroup', 'logs:PutLogEvents'],
        }),
      ],
    });

    return new iam.Role(this, 'ServiceRole', {
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      inlinePolicies: {
        canaryPolicy: policy,
      },
    });
  }

  /**
   * Returns the code object taken in by the canary resource.
   */
  private createCode(props: CanaryProps): CfnCanary.CodeProperty {
    const codeConfig = {
      handler: props.test.handler,
      ...props.test.code.bind(this, props.test.handler),
    };
    return {
      handler: codeConfig.handler,
      script: codeConfig.inlineCode,
      s3Bucket: codeConfig.s3Location?.bucketName,
      s3Key: codeConfig.s3Location?.objectKey,
      s3ObjectVersion: codeConfig.s3Location?.objectVersion,
    };
  }

  /**
   * Returns a canary schedule object
   */
  private createSchedule(props: CanaryProps): CfnCanary.ScheduleProperty {
    return {
      durationInSeconds: String(`${props.timeToLive?.toSeconds() ?? 0}`),
      expression: props.schedule?.expressionString ?? 'rate(5 minutes)',
    };
  }

  private createRunConfig(props: CanaryProps): CfnCanary.RunConfigProperty | undefined {
    if (!props.environmentVariables) {
      return undefined;
    }
    return {
      environmentVariables: props.environmentVariables,
    };
  }

  /**
   * Creates a unique name for the canary. The generated name is the physical ID of the canary.
   */
  private generateUniqueName(): string {
    const name = cdk.Names.uniqueId(this).toLowerCase().replace(' ', '-');
    if (name.length <= 21) {
      return name;
    } else {
      return name.substring(0, 15) + nameHash(name);
    }
  }

  private cannedMetric(
    fn: (dims: { CanaryName: string }) => MetricProps,
    props?: MetricOptions): Metric {
    return new Metric({
      ...fn({ CanaryName: this.canaryName }),
      ...props,
    }).attachTo(this);
  }
}

/**
 * Take a hash of the given name.
 *
 * @param name the name to be hashed
 */
function nameHash(name: string): string {
  const md5 = crypto.createHash('sha256').update(name).digest('hex');
  return md5.slice(0, 6);
}

const nameRegex: RegExp = /^[0-9a-z_\-]+$/;

/**
 * Verifies that the name fits the regex expression: ^[0-9a-z_\-]+$.
 *
 * @param name - the given name of the canary
 */
function validateName(name: string) {
  if (name.length > 21) {
    throw new Error(`Canary name is too large, must be between 1 and 21 characters, but is ${name.length} (got "${name}")`);
  }
  if (!nameRegex.test(name)) {
    throw new Error(`Canary name must be lowercase, numbers, hyphens, or underscores (got "${name}")`);
  }
}
