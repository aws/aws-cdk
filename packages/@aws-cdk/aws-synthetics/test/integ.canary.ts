/// !cdk-integ canary-one

import * as path from 'path';
import * as s3 from '@aws-cdk/aws-s3';
import * as cdk from '@aws-cdk/core';
import * as synthetics from '../lib';

/*
 * Stack verification steps:
 *
 * -- aws synthetics get-canary --name canary-integ has a state of 'RUNNING'
 * -- aws synthetics get-canary --name assetcanary-one has a state of 'RUNNING'
 * -- aws synthetics get-canary --name assetcanary-two has a state of 'RUNNING'
 */
const app = new cdk.App();
const stack = new cdk.Stack(app, 'canary-one');

const bucket = new s3.Bucket(stack, 'mytestbucket');
const prefix = 'integ';

new synthetics.Canary(stack, 'MyCanary', {
  canaryName: 'canary-integ',
  test: synthetics.Test.custom({
    handler: 'index.handler',
    code: synthetics.Code.fromInline(`
      exports.handler = async () => {
        console.log(\'hello world\');
      };`),
  }),
  schedule: synthetics.Schedule.rate(cdk.Duration.minutes(1)),
  artifactsBucketLocation: { bucket, prefix },
  runtime: synthetics.Runtime.SYNTHETICS_NODEJS_2_0,
});

new synthetics.Canary(stack, 'MyCanaryOne', {
  canaryName: 'assetcanary-one',
  test: synthetics.Test.custom({
    handler: 'canary.handler',
    code: synthetics.Code.fromAsset(path.join(__dirname, 'canaries')),
  }),
  runtime: synthetics.Runtime.SYNTHETICS_NODEJS_2_0,
});

new synthetics.Canary(stack, 'MyCanaryTwo', {
  canaryName: 'assetcanary-two',
  test: synthetics.Test.custom({
    handler: 'canary.handler',
    code: synthetics.Code.fromAsset(path.join(__dirname, 'canary.zip')),
  }),
  runtime: synthetics.Runtime.SYNTHETICS_NODEJS_2_0,
});

app.synth();
