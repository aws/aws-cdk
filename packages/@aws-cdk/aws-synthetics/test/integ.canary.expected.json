{
  "Resources": {
    "mytestbucket8DC16178": {
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
      "DeletionPolicy": "Retain"
    },
    "MyCanaryServiceRole593F9DD9": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "s3:ListAllMyBuckets",
                  "Effect": "Allow",
                  "Resource": "*"
                },
                {
                  "Action": [
                    "s3:PutObject",
                    "s3:GetBucketLocation"
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        {
                          "Fn::GetAtt": [
                            "mytestbucket8DC16178",
                            "Arn"
                          ]
                        },
                        "/integ/*"
                      ]
                    ]
                  }
                },
                {
                  "Action": "cloudwatch:PutMetricData",
                  "Condition": {
                    "StringEquals": {
                      "cloudwatch:namespace": "CloudWatchSynthetics"
                    }
                  },
                  "Effect": "Allow",
                  "Resource": "*"
                },
                {
                  "Action": [
                    "logs:CreateLogStream",
                    "logs:CreateLogGroup",
                    "logs:PutLogEvents"
                  ],
                  "Effect": "Allow",
                  "Resource": "arn:aws:logs:::*"
                }
              ],
              "Version": "2012-10-17"
            },
            "PolicyName": "canaryPolicy"
          }
        ]
      }
    },
    "MyCanary1A94CAFA": {
      "Type": "AWS::Synthetics::Canary",
      "Properties": {
        "ArtifactS3Location": {
          "Fn::Join": [
            "",
            [
              "s3://",
              {
                "Ref": "mytestbucket8DC16178"
              },
              "/integ"
            ]
          ]
        },
        "Code": {
          "Handler": "index.handler",
          "Script": "\n      exports.handler = async () => {\n        console.log('hello world');\n      };"
        },
        "ExecutionRoleArn": {
          "Fn::GetAtt": [
            "MyCanaryServiceRole593F9DD9",
            "Arn"
          ]
        },
        "Name": "canary-integ",
        "RuntimeVersion": "syn-1.0",
        "Schedule": {
          "DurationInSeconds": "0",
          "Expression": "rate(1 minute)"
        },
        "StartCanaryAfterCreation": true
      }
    }
  }
}